database-plugins/src/main/java/co/cask/hydrator/plugin/db/batch/source/DBSource.java

```diff
@@ -66,6 +83,127 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    public String importQuery;
+  }
+
+  @Path("listTables")
+  public List<String> listTables(ListTablesRequest request,
+                                 EndpointPluginContext pluginContext) throws SQLException,
+    InstantiationException, IllegalAccessException {
+
+    JDBCDriverShim driverShim = loadPluginClassAndGetDriver(request, pluginContext);
+    List<String> tableNames = new ArrayList();
+
+    try {
+      // initialize connection
+      Connection connection = getConnection(request.connectionString, request.user, request.password);
+      try {
+        DatabaseMetaData metadata = connection.getMetaData();
+        try (ResultSet rs = metadata.getTables(null, null, "%", null)) {
+          while (rs.next()) {
+            // table name is the third column and column index starts from 1.
+            String tableName = rs.getString(3);
+            if (tableName != null) {
+              tableNames.add(tableName);
+            }
+          }
+        }
+      } finally {
+        connection.close();
+      }
+      return tableNames;
+    } catch (SQLException e) {
+      LOG.error("Exception while trying to get the tables in a database for connection {}.",
+                request.connectionString, e);
+      throw Throwables.propagate(e);
+    } finally {
+      DBUtils.deregisterShim(driverShim);
+    }
+  }
+
+  private JDBCDriverShim loadPluginClassAndGetDriver(ListTablesRequest request, EndpointPluginContext pluginContext)
+    throws IllegalAccessException, InstantiationException, SQLException {
+    Class<? extends Driver> driverClass =
+      pluginContext.loadPluginClass(request.jdbcPluginType, request.jdbcPluginName, PluginProperties.builder().build());
+
+    try {
+      return DBUtils.ensureJDBCDriverIsAvailable(driverClass, request.connectionString,
+                                                 request.jdbcPluginType, request.jdbcPluginName);
+    } catch (IllegalAccessException | InstantiationException | SQLException e) {
+      LOG.error("Unable to load or register JDBC driver {} while getting the tables in a database.",
+                driverClass, e);
+      throw Throwables.propagate(e);
+    }
+  }
+
+  private Connection getConnection(String connectionString, String user, String password) throws SQLException {
+    if (user == null) {
+      return DriverManager.getConnection(connectionString);
+    } else {
+      return DriverManager.getConnection(connectionString, user, password);
+    }
+  }
+
+  @Path("getSchema")
+  public Map<String, String> getSchema(GetSchemaRequest request,
+                                       EndpointPluginContext pluginContext)
+    throws SQLException, InstantiationException, IllegalAccessException {
+
+    JDBCDriverShim driverShim = loadPluginClassAndGetDriver(request, pluginContext);
+
+    try {
+      Connection connection = getConnection(request.connectionString, request.user, request.password);
+      Map<String, String> resultSchema = new HashMap<>();
```

Comment:
Rather then using a `HashMap`, it's better to use a `LinkedHashMap` so that the ordering of the insertion is maintained (which is the ordering of the fields in the table).